android/KMAPro/kMAPro/src/main/java/com/keyman/android/CheckInstallReferrer.java

```diff
@@ -0,0 +1,147 @@
+            Log.i(TAG, "onInstallReferrerSetupFinished:OK");
+            processGooglePlayInstallReferrerData();
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
+            // API not available on the current Play Store app; log to Sentry
+            KMLog.LogError(TAG, "onInstallReferrerSetupFinished:FEATURE_NOT_SUPPORTED");
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE:
+            // Connection couldn't be established; log to Sentry
+            KMLog.LogError(TAG, "onInstallReferrerSetupFinished:SERVICE_UNAVAILABLE");
+            break;
+          default:
+            // There are some other error codes; log to Sentry
+            KMLog.LogError(TAG, "onInstallReferrerSetupFinished: Unexpected code: " +
+              responseCode);
+            break;
+        }
+        // We have finished with the API so tidy up
+        referrerClient.endConnection();
+      }
+
+      @Override
+      public void onInstallReferrerServiceDisconnected() {
+        // Try to restart the connection on the next request to
+        // Google Play by calling the startConnection() method.
+
+        // Note: the documentation on this is unclear, and examples I found typically have no
+        // implementation here, so for now, we'll just log this to Sentry so that we can be aware
+        // of it happening.
+        KMLog.LogError(TAG, "onInstallReferrerServiceDisconnected");
+      }
+
+      private void processGooglePlayInstallReferrerData() {
+        ReferrerDetails response;
+        try {
+          response = referrerClient.getInstallReferrer();
+        } catch (RemoteException e) {
+          KMLog.LogException(TAG, "Failed to get install referrer", e);
+          return;
+        }
+
+        String referrerUrl = response.getInstallReferrer();
+        if(referrerUrl == null) {
+          KMLog.LogError(TAG, "getInstallReferrer() returned null");
+          return;
+        }
+
+        Log.i(TAG, "Referrer URL: "+referrerUrl);
+        installPackageFromInstallReferrer(mainActivity, referrerUrl);
+      }
+    });
+  }
+
+  private static void installPackageFromInstallReferrer(MainActivity mainActivity,
+                                                        String urlReferrer) {
+    // We're going to try to and parse the string; if it looks like a valid referrer then we'll
+    // try and install the package. A valid referrer looks like:
+    //   source=keyman&package=<package>[&bcp47=<code>]
+    // Note that this will be encoded into a referrer url like this (i.e. = and & must be
+    // double-encoded to fit into the referrer parameter):
+    //   https://play.google.com/store/apps/details?id=com.tavultesoft.kmapro&
+    //   referrer=source%3Dkeyman%26package%3D<package>%26bcp47%3D<code>
+    // so we can use a standard URI parser to extract the parameters
+    Uri referrerUri = Uri.parse("https://example?" + urlReferrer);
```

Comment:
Looks like we only care about the query parameters. Is that why the URI can start with `https://example` (essentially anything)?

Reply:
Correct. We've just leveraging the existing url query param parser to avoid rolling our own parsing.